config/config.go

```diff
@@ -0,0 +1,343 @@
+	IPFSConn   jsonSection      `json:"ipfs_connector,omitempty"`
+	State      jsonSection      `json:"state,omitempty"`
+	PinTracker jsonSection      `json:"pin_tracker,omitempty"`
+	Monitor    jsonSection      `json:"monitor,omitempty"`
+	Allocator  jsonSection      `json:"allocator,omitempty"`
+	Informer   jsonSection      `json:"informer,omitempty"`
+}
+
+// Default generates a default configuration by generating defaults for all
+// registered components.
+func (cfg *Manager) Default() error {
+	for _, section := range cfg.sections {
+		for k, compcfg := range section {
+			logger.Debugf("generating default conf for %s", k)
+			err := compcfg.Default()
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// RegisterComponent lets the Manager load and save component configurations
+func (cfg *Manager) RegisterComponent(t ComponentType, ccfg ComponentConfig) error {
+	if cfg.sections == nil {
+		cfg.sections = make(map[ComponentType]Section)
+	}
+
+	_, ok := cfg.sections[t]
+	if !ok {
+		cfg.sections[t] = make(Section)
+	}
+
+	key := ccfg.ConfigKey()
+
+	cfg.sections[t][ccfg.ConfigKey()] = ccfg
+	if t == Cluster { // this is a special section
+		cfg.clusterKey = key
+	}
+
+	go cfg.watchSave(ccfg.SaveCh())
+	return nil
+}
+
+// Validate checks that all the registered componenets in this
+// Manager have valid configurations. It also makes sure that
+// the main Cluster compoenent exists.
+func (cfg *Manager) Validate() error {
+	if cfg.sections == nil {
+		return errors.New("no registered components")
+	}
+
+	if len(cfg.sections[Cluster]) < 1 {
+		return errors.New("no cluster component configuration")
+	}
+
+	for t, section := range cfg.sections {
+		if section == nil {
+			return fmt.Errorf("section %d is nil", t)
+		}
+		for k, compCfg := range section {
+			if compCfg == nil {
+				return fmt.Errorf("%s entry for section %d is nil", k, t)
+			}
+			err := compCfg.Validate()
+			if err != nil {
+				return fmt.Errorf("%s failed to validate: %s", k, err)
+			}
+		}
+	}
+	return nil
+}
+
+// LoadJSONFromFile reads a Configuration file from disk and parses
+// it. See LoadJSON too.
+func (cfg *Manager) LoadJSONFromFile(path string) error {
+	cfg.path = path
+
+	file, err := ioutil.ReadFile(path)
+	if err != nil {
+		logger.Error("error reading the configuration file: ", err)
+		return err
+	}
+
+	err = cfg.LoadJSON(file)
+	return err
+}
+
+// LoadJSON reads parses configurations for all registered components,
+// In order to work, component configurations must have been registered
+// beforehand with RegisterComponent.
+func (cfg *Manager) LoadJSON(bs []byte) error {
+	dir := filepath.Dir(cfg.path)
+
+	jcfg := &jsonConfig{}
+	err := json.Unmarshal(bs, jcfg)
+	if err != nil {
+		logger.Error("error parsing JSON: ", err)
+		return err
+	}
+
+	cfg.jsonCfg = jcfg
+
+	// Load Cluster section. Needs to have been registered
+	clustercfgs := cfg.sections[Cluster]
+	if clustercfgs != nil && len(clustercfgs) > 0 {
```

Comment:
yeah. would you prevent it?

Reply:
If you decide not to go with an explicit clusterSection in the Manager I think we should do something if `len(clustercfgs) > 1`, maybe an error, maybe a warning, because that would indicate that the user did not understand the invariant that there can only be one cluster section.